### ISSUE TITLE
bug in rule `font-weight-notation` in shorthand property `font`

### ISSUE THREAD

**Issue by evilebottnawi** [stylelint/stylelint#756]
Shorthand property `font` may any order property `font-style`, `font-weight`, `font-stretch`.
Example
```css
p.test {
    font: italic small-caps bold large serif;
}

p.test1 {
    font: small-caps italic bold large serif;
}

p.test2 {
    font: bold small-caps italic large serif;
}
...
```
But if i use and set `font-weight-notation` to `numeric`
```css
p {
    font: normal normal normal large serif;
}
```
I get error (it is normal):
```
2:11	⚠  Expected numeric font-weight notation (font-weight-notation) [stylelint]
```
I change code to:
```
p {
    font: normal normal 400 large serif;
}
```
But i get error:
```
2:11	⚠  Expected numeric font-weight notation (font-weight-notation) [stylelint]
```
Chrome opens all property to
```css
font-style: normal;
font-variant: normal;
font-weight: 400;
font-stretch: normal;
font-size: large;
line-height: normal;
font-family: serif;
```
I can change order, but get error all the same:
```css
p {
    font: 400 normal normal large serif;
}
```
```css
p {
    font: normal 400 normal large serif;
}
```

**Comment by ntwb** [comment]
Via https://developer.mozilla.org/en-US/docs/Web/CSS/font

> Note: There are a few caveats when using the CSS font shorthand. If these conditions are not met, the property is invalid and is entirely ignored.
> 1. Except when using a keyword, it is mandatory to define the value of both the `font-size` and the `font-family` properties.
> 2. Not all values of `font-variant` are allowed. Only the values defined in CSS 2.1 are allowed, that is `normal` and `small-caps`.
> 3. Though not directly settable by font, the values of `font-stretch`, `font-size-adjust` and `font-kerning` are also reset to their initial values.
> 4. The order of the values is not completely free: `font-style`, `font-variant` and `font-weight` must be defined, if any, before the `font-size` value. The `line-height` value must be defined immediately after the `font-size`, preceded by a mandatory /. Finally the `font-family` is mandatory and must be the last value defined (inherit value does not work).

A caveat on `font-stretch`: 
> The `font-stretch` property is new in CSS3 so if it is used in an older browser that doesn't support font-stretch in font shorthand, it will cause the entire line to be ignored.

As such IE, Edge, Firefox and Chrome support `font-stretch`, whereas Safari, Opera, iOS Safari, Opera Mini, Android and Chrome for Android do **not** support `font-stretch`

In my mind I see a stylelint rule for value order for the `font` shorthand property in order here, thinking about the regex required though is making my head explode :confounded:

**Comment by davidtheclark** [comment]
After reading a few things I still do not understand how this rule is supposed to work with those three initial values that each *could* have `normal` specified: style, variant, and weight. 

What does it mean if you have `font: normal 14px serif`? Is that `normal` style, variant, or weight? 

Or in `font: normal normal 14px serif` which one is *not* specified: style, variant, or weight? 

Additionally, am I right in thinking that there's no good reason to use `normal` in these shorthands, because by using the shorthand at all each of those values will be reset to its "initial" value, and for each of the 3 that is `normal`? Isn't `font: normal normal normal 14px serif` doing nothing more than `font: 14px serif`?

There's some language in [the spec] that's interesting, regarding `p { font: normal small-caps 120%/120% fantasy }`: 

> The fifth rule sets the 'font-variant' ('small-caps'), the 'font-size' (120% of the parent's font), the 'line-height' (120% times the font size) and the 'font-family' ('fantasy'). It follows that the keyword 'normal' applies to the two remaining properties: 'font-style' and 'font-weight'.

So in that case, the stylelint rule could complain that you're using `normal` when you should be using `400`, but you could think "Wha? I'm using `normal` for `font-style`!"

Am I wrong about the above? If not, maybe we should ignore `normal` in this rule and create a rule that disallows you from using `normal` in `font` because it's actually pointless?

(Seems I must be missing something ...)

**Comment by jeddy3** [comment]
>  Isn't font: normal normal normal 14px serif doing nothing more than font: 14px serif?

I've been trying things out in web inspector and it does behave this way.

>  If not, maybe we should ignore normal in this rule and create a rule that disallows you from using normal in font because it's actually pointless?

This sounds like the right course of action. A rule than enforces rewriting their example of `p { font: normal small-caps 120%/120% fantasy }` to `p { font: small-caps 120%/120% fantasy }` as it does the same thing?

But I too feel like we're missing something here. Hopefully @evilebottnawi can shed some light on it.

**Comment by evilebottnawi** [comment]
@jeddy3 maybe see how browsers process this property (determining which property is responsible for what - `font-style`, `font-weight`, `font-stretch`)? And use this algorithm in rule for shorthand property `font`.

**Comment by davidtheclark** [comment]
We could also try this logic:

- If no numbery value is found
  - and no font-weight specific keyword is found (e.g. `bold`, `bolder`)
    - and a `normal` is found
      - complain

**Comment by evilebottnawi** [comment]
@jeddy3 @davidtheclark 
Offical spec:
```
[ [ <‘font-style’> || <font-variant-css21> || <‘font-weight’> || <‘font-stretch’> ]? <‘font-size’> [ / <‘line-height’> ]? <‘font-family’> ] | caption | icon | menu | message-box | small-caption | status-bar
where 
<font-variant-css21> = [normal | small-caps]
```
https://developer.mozilla.org/ru/docs/Web/CSS/font

Main function to resolve shorthand:
```c++
void StylePropertySerializer::appendFontLonghandValueIfNotNormal(CSSPropertyID propertyID, StringBuilder& result, String& commonValue) const
{
    int foundPropertyIndex = m_propertySet.findPropertyIndex(propertyID);
    if (foundPropertyIndex == -1)
        return;

    const CSSValue* val = m_propertySet.propertyAt(foundPropertyIndex).value();
    if (val->isPrimitiveValue() && toCSSPrimitiveValue(val)->getValueID() == CSSValueNormal) {
        commonValue = String();
        return;
    }

    char prefix = '\0';
    switch (propertyID) {
    case CSSPropertyFontStyle:
        break; // No prefix.
    case CSSPropertyFontFamily:
    case CSSPropertyFontStretch:
    case CSSPropertyFontVariant:
    case CSSPropertyFontWeight:
        prefix = ' ';
        break;
    case CSSPropertyLineHeight:
        prefix = '/';
        break;
    default:
        ASSERT_NOT_REACHED();
    }

    if (prefix && !result.isEmpty())
        result.append(prefix);
    String value = m_propertySet.propertyAt(foundPropertyIndex).value()->cssText();
    result.append(value);
    if (!commonValue.isNull() && commonValue != value)
        commonValue = String();
}

String StylePropertySerializer::fontValue() const
{
    int fontSizePropertyIndex = m_propertySet.findPropertyIndex(CSSPropertyFontSize);
    int fontFamilyPropertyIndex = m_propertySet.findPropertyIndex(CSSPropertyFontFamily);
    if (fontSizePropertyIndex == -1 || fontFamilyPropertyIndex == -1) // First: we see on `font-size` and `font-family`, because it is minimal required properties - see specs above
        return emptyString();

    PropertyValueForSerializer fontSizeProperty = m_propertySet.propertyAt(fontSizePropertyIndex);
    PropertyValueForSerializer fontFamilyProperty = m_propertySet.propertyAt(fontFamilyPropertyIndex);

    String commonValue = fontSizeProperty.value()->cssText();
    StringBuilder result;
    // Second: see on `font-style`, `font-variant`, `font-weight`, `font-stretch`
    appendFontLonghandValueIfNotNormal(CSSPropertyFontStyle, result, commonValue);
    appendFontLonghandValueIfNotNormal(CSSPropertyFontVariant, result, commonValue);
    appendFontLonghandValueIfNotNormal(CSSPropertyFontWeight, result, commonValue);
    appendFontLonghandValueIfNotNormal(CSSPropertyFontStretch, result, commonValue);
    if (!result.isEmpty()) // Third: Bad resolving all values, skipping
        result.append(' ');
    result.append(fontSizeProperty.value()->cssText());
    // Fourthly: see on lint-height
    appendFontLonghandValueIfNotNormal(CSSPropertyLineHeight, result, commonValue);
    if (!result.isEmpty())
        result.append(' ');
    result.append(fontFamilyProperty.value()->cssText());
    // Fifth: check in is initial or inherit properties
    if (isInitialOrInherit(commonValue))
        return commonValue;
    return result.toString();
}
```
https://github.com/ChromiumWebApps/blink/blob/master/Source/core/css/StylePropertySerializer.cpp#L448
See comment, i think we need use same logic in rule for shorthand `font`.

**Comment by davidtheclark** [comment]
@evilebottnawi The trouble with trying to use the Chromium logic is that the goal of the browser is not the same as the goal of this property. We're not trying to determine what the final rendered value of `font-weight` should be: we're trying to determine whether a `normal` keyword has been used that deliberately targets `font-weight`.

**Comment by evilebottnawi** [comment]
@davidtheclark 
```css
* {
    font-style: italic;
}

p {
    normal 15px serif;
}
```
Normal is `font-style` (as in chromium code sequence)
